File: src/Vector.ts
export class Vector {
    x: number
    y: number

    constructor(x: number, y: number) {
        this.x = x
        this.y = y
    }

    heading() {
        return Math.atan2(this.y, this.x)
    }

    add(a: number, b: number) {
        this.x += a
        this.y += b
        return this
    }

    mult(a: number) {
        if (a != 0) {
            this.x *= a
            this.y *= a
        }
        return this
    }

    mag() {
        return Math.sqrt(Math.pow(this.x, 2) + Math.pow(this.y, 2))
    }

    unit() {
        if (this.mag() != 0) {
            let length = this.mag()
            this.x = this.x / length
            this.y = this.y / length
        }
    }

    static sub(a: Vector, b: Vector) {
        return new Vector(a.x - b.x, a.y - b.y)
    }
}

export function newVector(a: number, b: number) {
    const vec = new Vector(a, b)
    vec.x = a
    vec.y = b
    return vec
}

File: csharp/Utils/Vector.cs
using System;

namespace SmartRace.Utils
{
    public class Vector
    {
        public double X { get; set; }
        public double Y { get; set; }

        public Vector(double x, double y)
        {
            X = x;
            Y = y;
        }

        public double Heading()
        {
            return Math.Atan2(Y, X);
        }

        public Vector Add(double x, double y)
        {
            X += x;
            Y += y;
            return this;
        }

        public Vector Mult(double scalar)
        {
            if (scalar != 0)
            {
                X *= scalar;
                Y *= scalar;
            }
            return this;
        }

        public double Mag()
        {
            return Math.Sqrt(X * X + Y * Y);
        }

        public Vector Unit()
        {
            double length = Mag();
            if (length != 0)
            {
                X = X / length;
                Y = Y / length;
            }
            return this;
        }

        public static Vector Sub(Vector a, Vector b)
        {
            return new Vector(a.X - b.X, a.Y - b.Y);
        }

        public static Vector Add(Vector a, Vector b)
        {
            return new Vector(a.X + b.X, a.Y + b.Y);
        }

        public override string ToString()
        {
            return $"Vector({X}, {Y})";
        }
    }

    public static class VectorFactory
    {
        public static Vector NewVector(double x, double y)
        {
            return new Vector(x, y);
        }
    }
}